hal/src/stm32f2xx/usb_hal.c

```diff
@@ -164,20 +176,34 @@ uint8_t USB_USART_Available_Data(void)
  *******************************************************************************/
 int32_t USB_USART_Receive_Data(uint8_t peek)
 {
-    if(USB_Rx_State == 1)
+    if (USB_USART_Available_Data() > 0)
     {
-        if(!peek && ((USB_Rx_length - USB_Rx_ptr) == 1))
-        {
-            USB_Rx_State = 0;
-
-            /* Prepare Out endpoint to receive next packet */
-            DCD_EP_PrepareRx(&USB_OTG_dev,
-                             CDC_OUT_EP,
-                             (uint8_t*)(USB_Rx_Buffer),
-                             CDC_DATA_OUT_PACKET_SIZE);
+        uint32_t tail = USB_Rx_Buffer_tail;
+        uint8_t data = USB_Rx_Buffer[USB_Rx_Buffer_tail];
+        if (!peek) {
+            tail++;
+            if (tail == USB_Rx_Buffer_length)
+                tail = 0;
+            USB_Rx_Buffer_tail = tail;
         }
+        return data;
+    }
 
-        return USB_Rx_Buffer[peek ? USB_Rx_ptr : USB_Rx_ptr++];
+    return -1;
+}
+
+/*******************************************************************************
+ * Function Name  : USB_USART_Available_Data_For_Write.
+ * Description    : Return the length of available space in TX buffer
+ * Input          : None.
+ * Return         : Length.
+ *******************************************************************************/
+int32_t USB_USART_Available_Data_For_Write(void)
+{
+    if (USB_USART_Connected())
+    {
```

Comment:
Similarly here - because `USB_Tx_Buffer_tail` is updated by the USB IRQ, it's best to capture that as a local variable first, then use it for comparison and computation to avoid an inconsistent computation.